src/py/pyodide/console.py

```diff
@@ -262,95 +331,133 @@ def _stdstreams_redirections_inner(self):
         finally:
             self._streams_redirected = False
 
-    def flush_all(self):
-        """Force stdout/stderr flush."""
-        with self.redirect_streams():
-            sys.stdout.flush()
-            sys.stderr.flush()
+    def runsource(self, source: str, filename: str = "<console>") -> RunSourceResult:
+        """Compile and run source code in the interpreter.
+
+        Returns
+        -------
+            ``("incomplete", None)``
+                The source string is incomplete.
 
-    def runsource(self, *args, **kwargs):
-        """Force streams redirection.
+            ``("syntax-error", message ꞉ str)``
+                The source had a syntax error. ``message` is the formatted error
+                message as returned by :any:`Console.formatsyntaxerror`.
 
-        Syntax errors are not thrown by :any:`_InteractiveConsole.runcode` but
-        here in :any:`_InteractiveConsole.runsource`. This is why we force
-        redirection here since doing twice is not an issue.
+            ``("complete", future ꞉ Future[RunCodeResult])``
+                The source was complete and is being run. The ``Future`` will be
+                resolved with the result of :any:`Console.runcode` when it is finished
+                running.
         """
+        try:
+            code = self._compile(source, filename, "single")
```

Comment:
There are some funky issues with inheritance since the subclass methods are typed differently than the superclass. I will try to fix and see what happens.

Reply:
Okay so `_Compile` unconditionally returns a `CodeRunner`, but `_CommandCompiler` definitely returns `Optional[CodeRunner]`.